src/workerd/jsg/ser-test.c++

```diff
@@ -0,0 +1,226 @@
+    }
+
+    int getI() {
+      return i;
+    }
+
+    JSG_RESOURCE_TYPE(Foo) {
+      JSG_READONLY_PROTOTYPE_PROPERTY(i, getI);
+    }
+
+    void serialize(jsg::Lock& js, jsg::Serializer& serializer) {
+      serializer.writeRawUint32(i);
+    }
+    static jsg::Ref<Foo> deserialize(Lock& js, SerializationTag tag, Deserializer& deserializer) {
+      KJ_ASSERT(tag == SerializationTag::FOO);
+
+      // Intentionally deserialize differently so we can detect it.
+      return jsg::alloc<Foo>(deserializer.readRawUint32() + 2);
+    }
+    JSG_SERIALIZABLE(SerializationTag::FOO);
+  };
+
+  struct Bar: public jsg::Object {
+    kj::String text;
+    Bar(kj::String text): text(kj::mv(text)) {}
+
+    static jsg::Ref<Bar> constructor(kj::String text) {
+      return jsg::alloc<Bar>(kj::mv(text));
+    }
+
+    kj::String getText() {
+      return kj::str(text);
+    }
+
+    JSG_RESOURCE_TYPE(Bar) {
+      JSG_READONLY_PROTOTYPE_PROPERTY(text, getText);
+    }
+
+    void serialize(jsg::Lock& js, jsg::Serializer& serializer) {
+      serializer.writeRawUint64(text.size());
+      serializer.writeRawBytes(text.asBytes());
+    }
+    static jsg::Ref<Bar> deserialize(Lock& js, SerializationTag tag, Deserializer& deserializer) {
+      KJ_ASSERT(tag == SerializationTag::BAR);
+
+      size_t size = deserializer.readRawUint64();
+      auto bytes = deserializer.readRawBytes(size);
+      // Intentionally deserialize differently so we can detect it.
+      return jsg::alloc<Bar>(kj::str(bytes.asChars(), '!'));
+    }
+    JSG_SERIALIZABLE(SerializationTag::BAR);
+  };
+
+  JsValue roundTrip(Lock& js, JsValue in) {
+    auto content = ({
+      Serializer ser(js);
+      ser.write(js, in);
+      ser.release();
+    });
+
+    auto result = ({
+      Deserializer deser(js, content);
+      deser.readValue(js);
+    });
+
+    // Save the last serialization off to the side.
+    lastSerializedData = kj::mv(content.data);
+
+    return result;
+  }
+
+  JSG_RESOURCE_TYPE(SerTestContext) {
+    JSG_NESTED_TYPE(Foo);
+    JSG_NESTED_TYPE(Bar);
+    JSG_METHOD(roundTrip);
+  }
+};
+JSG_DECLARE_ISOLATE_TYPE(SerTestIsolate, SerTestContext, SerTestContext::Foo, SerTestContext::Bar);
+
+// Define a whole second JSG isolate type that contains "updated" code where Bar no longer wraps
+// a string, it wraps an arbitrary value.
+struct SerTestContextV2: public ContextGlobalObject {
+  enum class SerializationTag {
+    FOO,
+    BAR_OLD,
+    BAR_V2
+  };
+
+  struct Bar: public jsg::Object {
+    JsRef<JsValue> val;
+    Bar(JsRef<JsValue> val): val(kj::mv(val)) {}
+
+    static jsg::Ref<Bar> constructor(JsRef<JsValue> val) {
+      return jsg::alloc<Bar>(kj::mv(val));
+    }
+
+    JsRef<JsValue> getVal(Lock& js) {
+      return val.addRef(js);
+    }
+
+    JSG_RESOURCE_TYPE(Bar) {
+      JSG_READONLY_PROTOTYPE_PROPERTY(val, getVal);
+    }
+
+    void serialize(jsg::Lock& js, jsg::Serializer& serializer) {
+      // V2 just writes a value!
```

Comment:
Nit: a test variation that throws an exception in the `serialize(...)` and `deserialize(...)` methods and validates those are propagated correctly would be helpful.

Reply:
1780da8f